# Task 6.1 Sourcing Open Data


## Project Name: Citibike New York Bike Rider's Rental Analysis


## 01. Importing Libraries

     Step 1.1: Importing all necessary and needed libraries


In [2]:
#Step 1.1: Importing Libraries for Citibike Dataset; 

import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy
import calendar


## 02. Importing Citibike Dataset

     Step 2.1: Creating a Folderpath for the Dataset
     
     Step 2.2: Importing the Citibike Dataset for Exploratory Analysis

In [3]:
#Step 2.1. Creating a Folderpath for Citibike Dataset; 

path ='/users/ibu/Documents/Bike Rental Analysis'

In [4]:
#Verifying Path; 

path

'/users/ibu/Documents/Bike Rental Analysis'

In [5]:
#Step 2.2. Importing 'Citibike' Dataset from 'Original Data' Folder;

df_Citibike = pd.read_csv(os.path.join(path, '02 Data', 'Original', 'Citibike_Dataset_2013.csv'), index_col = False)



## 03. Data Profiling & Integrity

<b>Step 1: Data Profiling</b>

     Step 1.1: Dataframe Dimension & Shape
     Step 1.2: Summary Statistics Prior to Dataset Cleanup
     Step 1.3: Checking Dataframe Column Details
     Step 1.4: Cheking Dataframe Variables
     Step 1.5: Checking Dataframe Data Types

<b>Step 2: Data Integrity</b>

     Step 2.1: Dataframe Uniqueness: Finding Duplicates
     Step 2.2: Checking for Dataframe Mixed-Data Types 
     Step 2.3: Dataframe Completeness: Finding Missing Values
     Step 2.4: Dataset Timeliness: Collected September 2013

<b>Step 3: Changes Performed</b>
     
     Step 3.1: Defined 'Gender' Type Characteristics: Created New Column 'Rider_Gender'
     Step 3.2: Deleted Unnecessary Columns; 'Birth_Year', 'Trip_ID' & 'Gender'
     Step 3.3: Ran Descriptive Statistics after Cleanup



### Step 1: Data Profiling

In [6]:
#Step 1.1a: Checking for Dataframe Shape; 

df_Citibike.shape

(50000, 18)

<b>Notes:</b> Original Dataframe has 50,000 records (rows) and 18 Columns in Total.


In [7]:
#Step 1.1b: Checking Dataframe Dimensions;

df_Citibike.head()

Trip_ID  Bike_ID Weekday  Start_Hour     Start_Time  Start_Station_ID  \
0  LnQzQk    16013     Mon          18   9/9/13 18:18               523   
1  IL9boN    15230     Thu          18  9/12/13 18:38               257   
2  46clGB    17942     Wed          19  9/18/13 19:44               479   
3  v7vdFt    19683     Sat          11  9/28/13 11:54               527   
4  VGBsb5    18024     Sat          18   9/7/13 18:08               521   

        Start_Station_Name  Start_Station_Latitude  Start_Station_Longitude  \
0          W 38 St & 8 Ave               40.754666               -73.991382   
1  Lispenard St & Broadway               40.719392               -74.002472   
2          9 Ave & W 45 St               40.760193               -73.991255   
3          E 33 St & 1 Ave               40.743156               -73.974347   
4          8 Ave & W 31 St               40.750450               -73.994811   

        End_Time  End_Station_ID     End_Station_Name  End_Station_Latitude  \
0   9/9/13 18:35             334      W 20 St & 7 Ave             40.742388   
1  9/12/13 18:48             236  St Marks Pl & 2 Ave             40.728419   
2  9/18/13 19:50             513     W 56 St & 10 Ave             40.768254   
3  9/28/13 12:03             441      E 52 St & 2 Ave             40.756014   
4   9/7/13 18:46             476      E 31 St & 3 Ave             40.743943   

   End_Station_Longitude  Trip_Duration   Member_Status  Birth_Year  Gender  
0             -73.997262            993      Subscriber      1968.0       2  
1             -73.987140            581      Subscriber      1983.0       1  
2             -73.988639            361      Subscriber      1989.0       1  
3             -73.967416            561      Subscriber      1988.0       2  
4             -73.979661           2296  Non-Subscriber         NaN       0

<b>Notes:</b> Noting that line numbered 4 highlighted 'Trip_Duration' as 2296 yet rental 'Start_Time' is 18:08PM and End_Time is 18:46PM. This is like 36 minutes. Not sure if 2296 is in seconds or microseconds.


In [8]:
#Step 1.2: Running Descriptive Statistics prior to clean up; 

df_Citibike.describe()

Bike_ID    Start_Hour  Start_Station_ID  Start_Station_Latitude  \
count  50000.000000  50000.000000      50000.000000            50000.000000   
mean   17615.269360     14.145240        443.321500               40.734170   
std     1675.407446      4.860541        356.559925                0.019911   
min    14556.000000      0.000000         72.000000               40.680342   
25%    16188.000000     10.000000        304.000000               40.720196   
50%    17584.000000     15.000000        402.000000               40.735877   
75%    19014.000000     18.000000        484.000000               40.750020   
max    20642.000000     23.000000       3002.000000               40.770513   

       Start_Station_Longitude  End_Station_ID  End_Station_Latitude  \
count             50000.000000    50000.000000          50000.000000   
mean                -73.991109      442.539700             40.733859   
std                   0.012555      355.756022              0.019885   
min                 -74.017134       72.000000             40.680342   
25%                 -74.000271      304.000000             40.720196   
50%                 -73.990765      402.000000             40.735354   
75%                 -73.981923      483.000000             40.749013   
max                 -73.950048     3002.000000             40.770513   

       End_Station_Longitude  Trip_Duration    Birth_Year        Gender  
count           50000.000000   50000.000000  43021.000000  50000.000000  
mean              -73.991351     838.982900   1975.627786      1.073540  
std                 0.012569     573.663997     11.089001      0.589389  
min               -74.017134      60.000000   1899.000000      0.000000  
25%               -74.001547     417.000000   1968.000000      1.000000  
50%               -73.991218     672.000000   1978.000000      1.000000  
75%               -73.982050    1112.000000   1984.000000      1.000000  
max               -73.950048    2697.000000   1997.000000      2.000000

<b>Notes:</b> Seems 'Trip_Duration' is in seconds or microseconds.


In [9]:
#Copying Descriptive Statistics to clipboard before Cleanup;

df_Citibike.describe().to_clipboard()

In [10]:
#Step 1.3: Checking Dataframe Columns Detail; 

df_Citibike.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 18 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Trip_ID                  50000 non-null  object 
 1   Bike_ID                  50000 non-null  int64  
 2   Weekday                  50000 non-null  object 
 3   Start_Hour               50000 non-null  int64  
 4   Start_Time               50000 non-null  object 
 5   Start_Station_ID         50000 non-null  int64  
 6   Start_Station_Name       50000 non-null  object 
 7   Start_Station_Latitude   50000 non-null  float64
 8   Start_Station_Longitude  50000 non-null  float64
 9   End_Time                 50000 non-null  object 
 10  End_Station_ID           50000 non-null  int64  
 11  End_Station_Name         50000 non-null  object 
 12  End_Station_Latitude     50000 non-null  float64
 13  End_Station_Longitude    50000 non-null  float64
 14  Trip_Duration         

In [11]:
#Step 1.4 & 1.5: Checking for Dataframe Variables & Data Types; 

df_Citibike.dtypes

Trip_ID                     object
Bike_ID                      int64
Weekday                     object
Start_Hour                   int64
Start_Time                  object
Start_Station_ID             int64
Start_Station_Name          object
Start_Station_Latitude     float64
Start_Station_Longitude    float64
End_Time                    object
End_Station_ID               int64
End_Station_Name            object
End_Station_Latitude       float64
End_Station_Longitude      float64
Trip_Duration                int64
Member_Status               object
Birth_Year                 float64
Gender                       int64
dtype: object

In [12]:
#Copying 'Data-Types' to clipboard;

df_Citibike.dtypes.to_clipboard()


### Step 2: Data Integrity


In [13]:
#Step 2.1: Dataframe Uniqueness: Finding Duplicates;

df_Citibike_Dups = df_Citibike[df_Citibike.duplicated()]

In [14]:
#Printing Duplicates; 

df_Citibike_Dups

Empty DataFrame
Columns: [Trip_ID, Bike_ID, Weekday, Start_Hour, Start_Time, Start_Station_ID, Start_Station_Name, Start_Station_Latitude, Start_Station_Longitude, End_Time, End_Station_ID, End_Station_Name, End_Station_Latitude, End_Station_Longitude, Trip_Duration, Member_Status, Birth_Year, Gender]
Index: []

<b>Finding:</b> Result shows no duplicates exist.


In [15]:
#Step 2.2: Checking for Mixed-Data Types; 

for col in df_Citibike.columns.tolist():
  weird = (df_Citibike[[col]].applymap(type) != df_Citibike[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_Citibike[weird]) > 0:
    print (col)

<b>Finding:</b> No Mixed-Data Types Found.


In [16]:
#Step 2.3: Checking for Missing Values;

df_Citibike.isnull().sum()

Trip_ID                       0
Bike_ID                       0
Weekday                       0
Start_Hour                    0
Start_Time                    0
Start_Station_ID              0
Start_Station_Name            0
Start_Station_Latitude        0
Start_Station_Longitude       0
End_Time                      0
End_Station_ID                0
End_Station_Name              0
End_Station_Latitude          0
End_Station_Longitude         0
Trip_Duration                 0
Member_Status                 0
Birth_Year                 6979
Gender                        0
dtype: int64

<b>Notes:</b> 

The Column 'Birth_Year' with missing values 6,979 (14%) will be dropped since dataframe is more than 75% okay to proceed with the analysis without Rider's Birth_Year.

<b>Actions Taken:</b> Dropped the 'Birth_Year' Column


In [17]:
#Creating New Dataframe for the Actual Missing Values under 'Birth_Year';

df_Citibike_nan = df_Citibike[df_Citibike['Birth_Year'].isnull() == True]

In [18]:
#Viewing the Actual Missing Values under 'Birth_Year';

df_Citibike_nan

Trip_ID  Bike_ID Weekday  Start_Hour     Start_Time  Start_Station_ID  \
4      VGBsb5    18024     Sat          18   9/7/13 18:08               521   
10     aZsCdK    15732     Sat          20   9/7/13 20:01               457   
17     VqhF3S    15651     Mon          18  9/30/13 18:32               363   
26     9yZrq6    17506     Sun          16   9/1/13 16:18               389   
46     t1KBiF    19206     Sun          22   9/1/13 22:49               426   
...       ...      ...     ...         ...            ...               ...   
49986  8wDcq8    19293     Sun          22  9/29/13 22:00               469   
49988  C0ivmQ    18118     Sun          12  9/22/13 12:19               352   
49993  cvgHF4    18458     Wed          19  9/25/13 19:01               284   
49995  Fb2bmC    17988     Sat          18  9/21/13 18:39               479   
49998  xxacSt    16763     Sat          13   9/7/13 13:25               271   

           Start_Station_Name  Start_Station_Latitude  \
4             8 Ave & W 31 St               40.750450   
10         Broadway & W 58 St               40.766953   
17             West Thames St               40.708347   
26        Broadway & Berry St               40.710446   
46      West St & Chambers St               40.717548   
...                       ...                     ...   
49986      Broadway & W 53 St               40.763441   
49988         W 56 St & 6 Ave               40.763406   
49993   Greenwich Ave & 8 Ave               40.739017   
49995         9 Ave & W 45 St               40.760193   
49998  Ashland Pl & Hanson Pl               40.685282   

       Start_Station_Longitude       End_Time  End_Station_ID  \
4                   -73.994811   9/7/13 18:46             476   
10                  -73.981693   9/7/13 20:23             512   
17                  -74.017134  9/30/13 19:01             514   
26                  -73.965251   9/1/13 16:26             389   
46                  -74.013221   9/1/13 23:06             247   
...                        ...            ...             ...   
49986               -73.982681  9/29/13 22:17             444   
49988               -73.977225  9/22/13 12:29             464   
49993               -74.002638  9/25/13 19:19             329   
49995               -73.991255  9/21/13 20:08             478   
49998               -73.978058   9/7/13 14:14             373   

                   End_Station_Name  End_Station_Latitude  \
4                   E 31 St & 3 Ave             40.743943   
10                  W 29 St & 9 Ave             40.750073   
17                 12 Ave & W 40 St             40.760875   
26              Broadway & Berry St             40.710446   
46           Perry St & Bleecker St             40.735354   
...                             ...                   ...   
49986            Broadway & W 24 St             40.742354   
49988               E 56 St & 3 Ave             40.759345   
49993     Greenwich St & N Moore St             40.720434   
49995              11 Ave & W 41 St             40.760301   
49998  Willoughby Ave & Walworth St             40.693317   

       End_Station_Longitude  Trip_Duration   Member_Status  Birth_Year  \
4                 -73.979661           2296  Non-Subscriber         NaN   
10                -73.998393           1313  Non-Subscriber         NaN   
17                -74.002777           1751  Non-Subscriber         NaN   
26                -73.965251            478  Non-Subscriber         NaN   
46                -74.004831           1016  Non-Subscriber         NaN   
...                      ...            ...             ...         ...   
49986             -73.989151           1029  Non-Subscriber         NaN   
49988             -73.967597            611  Non-Subscriber         NaN   
49993             -74.010206           1084  Non-Subscriber         NaN   
49995             -73.998842           2697  Non-Subscriber         NaN   
49998             -73.953820     

<b>Notes:</b> Missing 'Birth_Year' Records = 6,979 / Original Dataframe has 50,000 records + 18 Columns



### Step 3: Changes Performed
     

In [19]:
#Step 3.1: Defining 'Gender' Type Characteristics;

Rider_Gender = []

for value in df_Citibike["Gender"]:
    if value == 1:
        Rider_Gender.append("Male")
    elif value == 2:
        Rider_Gender.append("Female")
    else:
        Rider_Gender.append("Unknown")

In [20]:
df_Citibike['Rider_Gender'] = Rider_Gender

In [21]:
df_Citibike['Rider_Gender'].value_counts(dropna = False)

Male       32361
Female     10658
Unknown     6981
Name: Rider_Gender, dtype: int64

In [22]:
#Checking for New Column; 

df_Citibike.head()

Trip_ID  Bike_ID Weekday  Start_Hour     Start_Time  Start_Station_ID  \
0  LnQzQk    16013     Mon          18   9/9/13 18:18               523   
1  IL9boN    15230     Thu          18  9/12/13 18:38               257   
2  46clGB    17942     Wed          19  9/18/13 19:44               479   
3  v7vdFt    19683     Sat          11  9/28/13 11:54               527   
4  VGBsb5    18024     Sat          18   9/7/13 18:08               521   

        Start_Station_Name  Start_Station_Latitude  Start_Station_Longitude  \
0          W 38 St & 8 Ave               40.754666               -73.991382   
1  Lispenard St & Broadway               40.719392               -74.002472   
2          9 Ave & W 45 St               40.760193               -73.991255   
3          E 33 St & 1 Ave               40.743156               -73.974347   
4          8 Ave & W 31 St               40.750450               -73.994811   

        End_Time  End_Station_ID     End_Station_Name  End_Station_Latitude  \
0   9/9/13 18:35             334      W 20 St & 7 Ave             40.742388   
1  9/12/13 18:48             236  St Marks Pl & 2 Ave             40.728419   
2  9/18/13 19:50             513     W 56 St & 10 Ave             40.768254   
3  9/28/13 12:03             441      E 52 St & 2 Ave             40.756014   
4   9/7/13 18:46             476      E 31 St & 3 Ave             40.743943   

   End_Station_Longitude  Trip_Duration   Member_Status  Birth_Year  Gender  \
0             -73.997262            993      Subscriber      1968.0       2   
1             -73.987140            581      Subscriber      1983.0       1   
2             -73.988639            361      Subscriber      1989.0       1   
3             -73.967416            561      Subscriber      1988.0       2   
4             -73.979661           2296  Non-Subscriber         NaN       0   

  Rider_Gender  
0       Female  
1         Male  
2         Male  
3       Female  
4      Unknown

<b>New Column:</b> 'Rider_Gender' Added


In [23]:
#Step 3.2: Deleting Unnecessary Columns; 

    #'Trip_ID', 'Gender' and 'Birth_Year' will not be needed on this analysis

df_Citibike_Clean = df_Citibike.drop(columns = ['Trip_ID', 'Birth_Year', 'Gender'])

In [24]:
df_Citibike_Clean

Bike_ID Weekday  Start_Hour     Start_Time  Start_Station_ID  \
0        16013     Mon          18   9/9/13 18:18               523   
1        15230     Thu          18  9/12/13 18:38               257   
2        17942     Wed          19  9/18/13 19:44               479   
3        19683     Sat          11  9/28/13 11:54               527   
4        18024     Sat          18   9/7/13 18:08               521   
...        ...     ...         ...            ...               ...   
49995    17988     Sat          18  9/21/13 18:39               479   
49996    18458     Sat          17  9/14/13 17:32               261   
49997    15054     Sat          18   9/7/13 18:08               402   
49998    16763     Sat          13   9/7/13 13:25               271   
49999    16794     Mon          21  9/23/13 21:54               515   

            Start_Station_Name  Start_Station_Latitude  \
0              W 38 St & 8 Ave               40.754666   
1      Lispenard St & Broadway               40.719392   
2              9 Ave & W 45 St               40.760193   
3              E 33 St & 1 Ave               40.743156   
4              8 Ave & W 31 St               40.750450   
...                        ...                     ...   
49995          9 Ave & W 45 St               40.760193   
49996     Johnson St & Gold St               40.694749   
49997       Broadway & E 22 St               40.740343   
49998   Ashland Pl & Hanson Pl               40.685282   
49999         W 43 St & 10 Ave               40.760094   

       Start_Station_Longitude       End_Time  End_Station_ID  \
0                   -73.991382   9/9/13 18:35             334   
1                   -74.002472  9/12/13 18:48             236   
2                   -73.991255  9/18/13 19:50             513   
3                   -73.974347  9/28/13 12:03             441   
4                   -73.994811   9/7/13 18:46             476   
...                        ...            ...             ...   
49995               -73.991255  9/21/13 20:08             478   
49996               -73.983625  9/14/13 17:52             151   
49997               -73.989551   9/7/13 18:25             524   
49998               -73.978058   9/7/13 14:14             373   
49999               -73.994618  9/23/13 22:00             488   

                   End_Station_Name  End_Station_Latitude  \
0                   W 20 St & 7 Ave             40.742388   
1               St Marks Pl & 2 Ave             40.728419   
2                  W 56 St & 10 Ave             40.768254   
3                   E 52 St & 2 Ave             40.756014   
4                   E 31 St & 3 Ave             40.743943   
...                             ...                   ...   
49995              11 Ave & W 41 St             40.760301   
49996      Cleveland Pl & Spring St             40.721816   
49997               W 43 St & 6 Ave             40.755273   
49998  Willoughby Ave & Walworth St             40.693317   
49999               W 39 St & 9 Ave             40.756458   

       End_Station_Longitude  Trip_Duration   Member_Status Rider_Gender  
0                 -73.997262            993      Subscriber       Female  
1                 -73.987140            581      Subscriber         Male  
2                 -73.988639            361      Subscriber         Male  
3                 -73.967416            561      Subscriber       Female  
4                 -73.979661           2296  Non-Subscriber      Unknown  
...                      ...            ...             ...          ...  
49995             -73.998842           2697  Non-Subscriber      Unknown  
49996             -73.997203           1191      Subscriber       Female  
49997             -73.983169           1050      Subscriber       Female  
49998             -73.953820           2697  Non-Subscriber      Unknown  
49999             -73.993722            370      Subscriber       Female  

[50000 rows x 16 columns]

In [25]:
#Step 3.3: Running Descriptive Statistics after Cleanup; 

df_Citibike_Clean.describe()

Bike_ID    Start_Hour  Start_Station_ID  Start_Station_Latitude  \
count  50000.000000  50000.000000      50000.000000            50000.000000   
mean   17615.269360     14.145240        443.321500               40.734170   
std     1675.407446      4.860541        356.559925                0.019911   
min    14556.000000      0.000000         72.000000               40.680342   
25%    16188.000000     10.000000        304.000000               40.720196   
50%    17584.000000     15.000000        402.000000               40.735877   
75%    19014.000000     18.000000        484.000000               40.750020   
max    20642.000000     23.000000       3002.000000               40.770513   

       Start_Station_Longitude  End_Station_ID  End_Station_Latitude  \
count             50000.000000    50000.000000          50000.000000   
mean                -73.991109      442.539700             40.733859   
std                   0.012555      355.756022              0.019885   
min                 -74.017134       72.000000             40.680342   
25%                 -74.000271      304.000000             40.720196   
50%                 -73.990765      402.000000             40.735354   
75%                 -73.981923      483.000000             40.749013   
max                 -73.950048     3002.000000             40.770513   

       End_Station_Longitude  Trip_Duration  
count           50000.000000   50000.000000  
mean              -73.991351     838.982900  
std                 0.012569     573.663997  
min               -74.017134      60.000000  
25%               -74.001547     417.000000  
50%               -73.991218     672.000000  
75%               -73.982050    1112.000000  
max               -73.950048    2697.000000

In [26]:
#Copying Descriptive Statistics to clipboard After Cleanup;

df_Citibike_Clean.describe().to_clipboard()


## 04. Exporting Cleaned 'Citibike' Dataset


In [27]:
#Exporting Cleaned 'df_Citibike_Clean' Dataset;

df_Citibike_Clean.to_csv(os.path.join(path, '02 Data', 'Prepared', 'Citibike_Cleaned.csv'))